contracts/src/Epoch.sol

```diff
@@ -0,0 +1,140 @@
+        string workerPublicKey;
+        string networkKey;
+    }
+
+    /// epoch => committee public keys
+    mapping(uint256 => string[]) public committee;
+
+    function initialize(
+        address _nodeRegistry,
+        uint256 _firstEpochStart,
+        uint256 _epochDuration,
+        uint256 _maxCommitteeSize
+    ) external {
+        require(!initialized, "contract already initialized");
+        nodeRegistry = NodeRegistry(_nodeRegistry);
+        maxCommitteeSize = _maxCommitteeSize;
+        epochDurationMs = _epochDuration;
+        currentEpochEndStampMs = _firstEpochStart + _epochDuration;
+
+        committee[epoch] = _chooseNewCommittee();
+        currentCommitteeSize = committee[epoch].length;
+        initialized = true;
+    }
+
+    function getCurrentEpochInfo()
+        public
+        view
+        returns (uint256 _epoch, uint256 _currentEpochEndMs, CommitteeMember[] memory _committeeMembers)
+    {
+        _committeeMembers = new CommitteeMember[](currentCommitteeSize);
+        for (uint256 i; i < currentCommitteeSize;) {
+            NodeRegistry.Node memory node = nodeRegistry.getNodeInfo(committee[epoch][i]);
+            _committeeMembers[i].publicKey = committee[epoch][i];
+            _committeeMembers[i].primaryAddress = node.primaryAddress;
+            _committeeMembers[i].workerAddress = node.workerAddress;
+            _committeeMembers[i].workerPublicKey = node.workerPublicKey;
+            _committeeMembers[i].networkKey = node.networkKey;
+            _committeeMembers[i].workerMempool = node.workerMempool;
+
+            unchecked {
+                i += 1;
+            }
+        }
+
+        return (epoch, currentEpochEndStampMs, _committeeMembers);
+    }
+
+    function getCurrentCommittee() public view returns (string[] memory) {
+        return committee[epoch];
+    }
+
+    function signalEpochChange(string memory committeeMember) external returns (bool) {
+        for (uint256 i;;) {
+            if (keccak256(abi.encodePacked(committee[epoch][i])) == keccak256(abi.encodePacked(committeeMember))) {
+                readyToChange++;
+                break;
+            }
+            unchecked {
+                i++;
+            }
+            if (i == currentCommitteeSize) {
+                revert();
+            }
+        }
+        uint256 roundUp;
+        if (currentCommitteeSize * 2 / 3 > 0) {
+            roundUp = 1;
+        }
+
+        if (readyToChange >= (currentCommitteeSize * 2 / 3) + roundUp) {
+            _changeEpoch();
+            return (true);
+        } else {
+            return (false);
+        }
+    }
+
+    function _changeEpoch() private {
+        epoch++;
+
+        committee[epoch] = _chooseNewCommittee();
+        currentEpochEndStampMs = currentEpochEndStampMs + epochDurationMs;
+        currentCommitteeSize = committee[epoch].length;
+    }
+
+    function _chooseNewCommittee() private view returns (string[] memory _committee) {
```

Comment:
is the new committee drive by the application layer?

Reply:
Yes the new committee is decided by the application layer. But the process is only triggered by Narwhal certificates. The random function that selects the new committee is a placeholder for now. A truly random threshold signature that all nodes have a chance to contribute too is the end goal here.